File: src/main/java/com/demonwav/statcraft/listeners/FishCaughtListener.java
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.listeners;

import com.demonwav.statcraft.StatCraft;
import com.demonwav.statcraft.Util;
import com.demonwav.statcraft.magic.FishCode;
import com.demonwav.statcraft.querydsl.QFishCaught;

import org.bukkit.entity.Item;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerFishEvent;

import java.util.UUID;

public class FishCaughtListener implements Listener {

    private final StatCraft plugin;

    public FishCaughtListener(final StatCraft plugin) {
        this.plugin = plugin;
    }

    @SuppressWarnings("deprecation")
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onFishCatch(final PlayerFishEvent event) {
        if (event.getCaught() != null) {
            final UUID uuid = event.getPlayer().getUniqueId();
            final String worldName = event.getPlayer().getWorld().getName();
            // shouldn't be an issue, but just to protect against a ClassCastException
            if (event.getCaught() instanceof Item) {
                Item item = (Item) event.getCaught();
                final short itemId = (short) item.getItemStack().getTypeId();
                final short damage = Util.damageValue(itemId, item.getItemStack().getData().getData());

                final FishCode code;

                switch (item.getItemStack().getType()) {
                    case RAW_FISH:
                        code = FishCode.FISH;
                        break;
                    case BOW:
                    case ENCHANTED_BOOK:
                    case NAME_TAG:
                    case SADDLE:
                    case WATER_LILY:
                        code = FishCode.TREASURE;
                        break;
                    case BOWL:
                    case LEATHER:
                    case LEATHER_BOOTS:
                    case ROTTEN_FLESH:
                    case STICK:
                    case STRING:
                    case POTION:
                    case BONE:
                    case INK_SACK:
                    case TRIPWIRE_HOOK:
                        code = FishCode.JUNK;
                        break;
                    case FISHING_ROD:
                        if (item.getItemStack().getEnchantments().size() == 0)
                            code = FishCode.JUNK;
                        else
                            code = FishCode.TREASURE;
                        break;
                    default:
                        // Default to junk, though it should never fall under this category
                        code = FishCode.JUNK;
                        break;
                }

                plugin.getThreadManager().schedule(
                    QFishCaught.class, uuid, worldName,
                    (f, clause, id, worldId) ->
                        clause.columns(f.id, f.worldId, f.item, f.damage, f.type, f.amount)
                            .values(id, worldId, itemId, damage, code.getCode(), 1).execute(),
                    (f, clause, id, worldId) ->
                        clause.where(f.id.eq(id), f.worldId.eq(worldId), f.item.eq(itemId), f.damage.eq(damage), f.type.eq(code.getCode()))
                            .set(f.amount, f.amount.add(1)).execute()
                );
            }
        }
    }
}

File: src/main/kotlin/com/demonwav/statcraft/listeners/FishCaughtListener.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.listeners

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.Util
import com.demonwav.statcraft.magic.FishCode
import com.demonwav.statcraft.querydsl.QFishCaught
import org.bukkit.Material
import org.bukkit.entity.Item
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerFishEvent

class FishCaughtListener(private val plugin: StatCraft) : Listener {

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onFishCatch(event: PlayerFishEvent) {
        if (event.caught == null) {
            return
        }

        val uuid = event.player.uniqueId
        val worldName = event.player.world.name
        val caught = event.caught

        if (caught is Item) {
            val itemId = caught.itemStack.typeId.toShort()
            val damage = Util.damageValue(itemId, caught.itemStack.data.data.toShort())

            val code = when (caught.itemStack.type) {
                Material.RAW_FISH ->
                    FishCode.FISH

                Material.BOW,
                Material.ENCHANTED_BOOK,
                Material.NAME_TAG,
                Material.SADDLE,
                Material.WATER_LILY ->
                    FishCode.TREASURE

                Material.BOWL,
                Material.LEATHER,
                Material.LEATHER_BOOTS,
                Material.ROTTEN_FLESH,
                Material.STICK,
                Material.STRING,
                Material.POTION,
                Material.BONE,
                Material.INK_SACK,
                Material.TRIPWIRE_HOOK ->
                    FishCode.JUNK

                Material.FISHING_ROD ->
                    if (caught.itemStack.enchantments.size == 0) {
                        FishCode.JUNK
                    } else {
                        FishCode.TREASURE
                    }

                else ->
                    FishCode.JUNK
            }

            plugin.threadManager.schedule<QFishCaught>(
                uuid, worldName,
                { f, clause, id, worldId ->
                    clause.columns(f.id, f.worldId, f.item, f.damage, f.amount)
                        .values(id, worldId, itemId, damage, code.code, 1).execute()
                }, { f, clause, id, worldId ->
                    clause.where(f.id.eq(id), f.worldId.eq(worldId), f.item.eq(itemId), f.damage.eq(damage), f.type.eq(code.code))
                        .set(f.amount, f.amount.add(1)).execute()
                }
            )
        }
    }
}
